src/include/miopen/conv/asm_implicit_gemm.hpp

```diff
@@ -225,6 +225,43 @@ HeuristicInitMacroTileNoPadGemmK(size_t gemm_m,
         return std::make_tuple(0, 0, 0);
 }
 
+// this is to support big tensor > 4G. need to decide how many splits needed
+// return the number of splits
+static inline int igemm_split_batch_size(const int hi,
+                                         const int wi,
+                                         const int ho,
+                                         const int wo,
+                                         const int n,
+                                         const int k,
+                                         const int c,
+                                         const int data_byte)
+{
+    size_t image_size_input = static_cast<size_t>(c) * hi * wi * data_byte;
+    size_t image_size_output = static_cast<size_t>(k) * ho * wo * data_byte;
+    size_t size_4g = 0xffffffffUL;
+    if(image_size_input >= size_4g || image_size_output >= size_4g)
+        return 0;
+
+    size_t image_size = image_size_input >= image_size_output ? image_size_input : image_size_output;
+    size_t splited_n = size_4g / image_size;
+
+    // round up splits, we must match
+    // 1. splited_n * image_size < size_4g
+    // 2. n % splited_n == 0
+    // if(splited_n >= n)
+    //     return 1;
+    if(splited_n == 0)
+        return 0;
+    while(splited_n >= 1){
+        if(n % splited_n == 0 && splited_n * image_size < size_4g)
```

Comment:
When `splited_n == 1` this folds into
```
if(n % 1 == 0 && 1 * image_size < size_4g)
```
which is equivalent to
```
if(0 == 0 && true)
```
which means that:
- splited_n can't reach 0 in this loop, and thus lines 260-261 are dead code
- the loop header can be changed to `while(splited_n > 1)`

Reply:
Another problem.

- AFAIU we want to find the biggest value `splited_n` in the range `[1...(size_4g / image_size)]` so that `n % splited_n == 0`. 
- The biggest possible number is `(size_4g / image_size)`, if `n` is divisible by this number without remainder.
- However this algorithm will refuse such a number when `size_4g % image_size == 0`.

Why: in the case described above, `splited_n * image_size == size_4g` and the 2nd part of the following condition is not met:
```
if(n % splited_n == 0 && splited_n * image_size < size_4g)
```
Is `<` correct?

Reply:
> splited_n can't reach 0 in this loop, and thus lines 260-261 are dead code

Lines 260-261 are dead code according to max_n's possible value. But without them, clang-tidy keeps throwing divided by zero warning in ```return n/max_n```

> the loop header can be changed to while(splited_n > 1)

agreed.

Reply:
> Another problem.
> 
> * AFAIU we want to find the biggest value `splited_n` in the range `[1...(size_4g / image_size)]` so that `n % splited_n == 0`.
> * The biggest possible number is `(size_4g / image_size)`, if `n` is divisible by this number without remainder.
> * However this algorithm will refuse such a number when `size_4g % image_size == 0`.
> 
> Why: in the case described above, `splited_n * image_size == size_4g` and the 2nd part of the following condition is not met:
> 
> ```
> if(n % splited_n == 0 && splited_n * image_size < size_4g)
> ```
> 
> Is `<` correct?

Yes, ```splited_n * image_size < size_4g``` will never meet. Remove this condition.